### PR TITLE
KAFKA-6555: Making state store queryable during restoration

### PR DESCRIPTION
**Pull request by a-surana**
*State store in Kafka streams are currently only queryable when StreamTask is in RUNNING state. The idea is to make it queryable even in the REBALANCING state as the time spend on restoration can be huge and making the data inaccessible during this time could be downtime not suitable for many applications. [https://issues.apache.org/jira/browse/KAFKA-6555]*

*Existing unit tests/integration tests are running fine*

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

[unit_tests.html.zip](https://github.com/apache/kafka/files/1729175/unit_tests.html.zip)

### CODE REVIEW

**Comment by a-surana** [comment]
https://issues.apache.org/jira/browse/KAFKA-6555

This is to also expose state-stores of restoring task's along with the running task's for read access.

**Comment by dguy** [comment]
Thanks @asurana28 for the PR. I can understand being able to read from the running tasks stores, but I'm not so sure about the restoring tasks. They may be a long way behind, so you could be reading very old data. I'm not sure this is always desirable @mjsax, @guozhangwang, @bbejeck

**Comment by mjsax** [comment]
@dguy There was a discussion on the JIRA -- not sure, if we reached consensus what design we want to have. Feel free to comment on the JIRA.

**Comment by guozhangwang** [comment]
> @dguy There was a discussion on the JIRA -- not sure, if we reached consensus what design we want to have. Feel free to comment on the JIRA.

I also left a comment on the JIRA. To quick sum up I'm +1 on this approach AND to do KAFKA-6145 additionally, but we should add a config for it and default to turn it off.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/AssignmentInfo.java [hunk 1]
```diff
@@ -122,7 +124,7 @@ public int latestSupportedVersion() {
     public ByteBuffer encode() {
         final ByteArrayOutputStream baos = new ByteArrayOutputStream();
 
-        try (final DataOutputStream out = new DataOutputStream(baos)) {
+        try (final DataOutputStream out = new DataOutputStream(new GZIPOutputStream(baos))) {
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/AssignmentInfo.java:127]: How does this change related to the JIRA?

Comment (a-surana) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/AssignmentInfo.java:127]: My bad, this change was for https://issues.apache.org/jira/browse/KAFKA-7149, but somehow got added to this PR itself. Please ignore changes in this file.

**Comment by mjsax** [comment]
Closing this PR, because the corresponding ticket is closed already.

**Pull request closed without merging**